pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -296,15 +298,290 @@ func (m *kubeGenericRuntimeManager) GetPods(all bool) ([]*kubecontainer.Pod, err
+		sandboxStatus.Linux.Namespaces.Options.GetHostNetwork() != kubecontainer.IsHostNetworkPod(pod) {
+		return true, sandboxStatus.Metadata.GetAttempt(), ""
+	}
+
+	return false, sandboxStatus.Metadata.GetAttempt(), sandboxStatus.GetId()
+}
+
+func (m *kubeGenericRuntimeManager) computePodContainerChanges(pod *api.Pod, podStatus *kubecontainer.PodStatus) podContainerSpecChanges {
+	glog.V(5).Infof("Syncing Pod %q: %+v", format.Pod(pod), pod)
+
+	sandboxChanged, attempt, sandboxID := m.podSandboxChanged(pod, podStatus)
+	changes := podContainerSpecChanges{
+		CreateSandbox:     sandboxChanged,
+		SandboxID:         sandboxID,
+		Attempt:           attempt,
+		ContainersToStart: make(map[int]string),
+		ContainersToKeep:  make(map[kubecontainer.ContainerID]int),
+	}
+
+	for index, container := range pod.Spec.Containers {
+		if sandboxChanged {
+			message := fmt.Sprintf("Container %+v's pod sandbox is dead, the container will be recreated.", container)
+			glog.Info(message)
+			changes.ContainersToStart[index] = message
+			continue
+		}
+
+		containerStatus := podStatus.FindContainerStatusByName(container.Name)
+		if containerStatus == nil || containerStatus.State != kubecontainer.ContainerStateRunning {
+			if kubecontainer.ShouldContainerBeRestarted(&container, pod, podStatus) {
+				message := fmt.Sprintf("Container %+v is dead, but RestartPolicy says that we should restart it.", container)
+				glog.Info(message)
+				changes.ContainersToStart[index] = message
+			}
+			continue
+		}
+
+		expectedHash := kubecontainer.HashContainer(&container)
+		containerChanged := containerStatus.Hash != 0 && containerStatus.Hash != expectedHash
+		if containerChanged {
+			message := fmt.Sprintf("Pod %q container %q hash changed (%d vs %d), it will be killed and re-created.",
+				pod.Name, container.Name, containerStatus.Hash, expectedHash)
+			glog.Info(message)
+			changes.ContainersToStart[index] = message
+			continue
+		}
+
+		liveness, found := m.livenessManager.Get(containerStatus.ID)
+		if !found || liveness == proberesults.Success {
+			changes.ContainersToKeep[containerStatus.ID] = index
+			continue
+		}
+		if pod.Spec.RestartPolicy != api.RestartPolicyNever {
+			message := fmt.Sprintf("pod %q container %q is unhealthy, it will be killed and re-created.", format.Pod(pod), container.Name)
+			glog.Info(message)
+			changes.ContainersToStart[index] = message
+		}
+	}
+
+	return changes
+}
+
 // SyncPod syncs the running pod into the desired pod.
-func (m *kubeGenericRuntimeManager) SyncPod(pod *api.Pod, _ api.PodStatus,
-	podStatus *kubecontainer.PodStatus, pullSecrets []api.Secret,
-	backOff *flowcontrol.Backoff) (result kubecontainer.PodSyncResult) {
-	result.Fail(fmt.Errorf("not implemented"))
+// TODO: support init containers in SyncPod.
+func (m *kubeGenericRuntimeManager) SyncPod(pod *api.Pod, _ api.PodStatus, podStatus *kubecontainer.PodStatus, pullSecrets []api.Secret, backOff *flowcontrol.Backoff) (result kubecontainer.PodSyncResult) {
```

Comment:
My first suggestion is to add a doc to express what the function is actually doing in steps.
Then probably split the function into steps.
Otherwise it's really hard for readers to reason about what's going on here..